src/core/CorePool.ts

```diff
@@ -311,11 +343,63 @@ export class CorePool {
     tickUpper: number,
     liquidityDelta: JSBI
   ): { position: Position; amount0: JSBI; amount1: JSBI } {
-    // TODO
+    let amount0: JSBI = ZERO,
+      amount1: JSBI = ZERO;
+
+    let position: Position = this.getPosition(owner, tickLower, tickUpper);
+
+    // use assert
+    if (tickLower > tickUpper) {
+      // error handling
+      console.error("[Error]: tickLower upper than tickUpper");
+    } else if (tickLower < TickMath.MIN_TICK) {
+      console.error("[Error]: tickLower lower than MIN_TICK");
+    } else if (tickUpper > TickMath.MAX_TICK) {
+      console.error("[Error]: tickUpper bigger than MAX_TICK");
+    } else {
+      // check ticks pass, update position
+      this.updatePosition(owner, tickLower, tickUpper, liquidityDelta);
+      // use switch or pattern matching
+      // check if liquidity happen add() or remove()
+      if (JSBI.notEqual(liquidityDelta, ZERO)) {
+        if (this._tickCurrent < tickLower) {
+          amount0 = SqrtPriceMath.getAmount0Delta(
+            TickMath.getSqrtRatioAtTick(tickLower),
+            TickMath.getSqrtRatioAtTick(tickUpper),
+            liquidityDelta
+          );
+        } else if (tickLower < this._tickCurrent) {
+          amount0 = SqrtPriceMath.getAmount0Delta(
+            TickMath.getSqrtRatioAtTick(tickLower),
+            TickMath.getSqrtRatioAtTick(tickUpper),
+            liquidityDelta
+          );
+
+          amount1 = SqrtPriceMath.getAmount1Delta(
+            TickMath.getSqrtRatioAtTick(tickLower),
+            TickMath.getSqrtRatioAtTick(tickUpper),
```

Comment:
The same above.
`TickMath.getSqrtRatioAtTick(tickUpper)` -> `this.sqrtPriceX96`